rs/cli/src/commands/mod.rs

```diff
@@ -48,24 +51,90 @@ pub mod upgrade;
 mod version;
 mod vote;
 
-#[derive(Parser, Debug)]
-#[clap(version = env!("CARGO_PKG_VERSION"), about, author)]
-pub struct Args {
+// Stupid clap does not let me define this
+// and the below struct as enums.
+// https://github.com/clap-rs/clap/issues/2621
+/// HSM authentication arguments
+#[derive(ClapArgs, Debug, Clone)]
+pub struct HsmOpts {
     /// Pin for the HSM key used for submitting proposals
-    #[clap(long, global = true, hide_env_values = true, env = "HSM_PIN")]
-    pub hsm_pin: Option<String>,
+    #[clap(
+        required = false,
+        alias = "hsm-pim",
+        requires_all = ["hsm_pin", "hsm_slot","hsm_key_id"],
+        conflicts_with = "private_key_pem",
+        long,
+        global = true,
+        hide_env_values = true,
+        env = "HSM_PIN"
+    )]
+    pub hsm_pin: String,
 
     /// Slot that HSM key uses, can be read with pkcs11-tool
-    #[clap(long, value_parser=maybe_hex::<u64>, global = true, env = "HSM_SLOT")]
-    pub hsm_slot: Option<u64>,
+    #[clap(required = false,
+        requires_all = ["hsm_pin", "hsm_slot","hsm_key_id"],
+        conflicts_with = "private_key_pem",
+        long, value_parser=maybe_hex::<u64>, global = true, env = "HSM_SLOT")]
+    pub hsm_slot: u64,
 
     /// HSM Key ID, can be read with pkcs11-tool
-    #[clap(long, global = true, env = "HSM_KEY_ID")]
-    pub hsm_key_id: Option<String>,
+    #[clap(
+        required = false,
+        requires_all = ["hsm_pin", "hsm_slot","hsm_key_id"],
+        conflicts_with = "private_key_pem",
+        long,
+        global = true,
+        env = "HSM_KEY_ID"
+    )]
+    pub hsm_key_id: String,
+}
 
-    /// Path to key pem file
-    #[clap(long, global = true, env = "PRIVATE_KEY_PEM")]
-    pub private_key_pem: Option<PathBuf>,
+// Stupid clap does not let me define this
+// and the above struct as enums.
+// https://github.com/clap-rs/clap/issues/2621
```

Comment:
Understand the frustration, but maybe not leave it like this in a public repo 😟

Reply:
Fair.

Reply:
This has been addressed.  PTAL @NikolaMilosa .